### ISSUE TITLE
Not working with PongDeterministic-v4

### ISSUE THREAD

**Issue by guptavaibhav18197** [keras-rl/keras-rl#213]
I tried running the code for PongDeterministic-v4 for 1.75 M steps but there was no improvement in score. It started with avg -21 but even after 1.75 M step the avg score was around -20.4. In the last testing phase too - the score was -21.0 throughout.

Any help with issue will be really appreciated. 

Thanks

**Comment by mirraaj** [comment]
Which algorithm did you try for PongDeterministic-v4?

**Comment by guptavaibhav18197** [comment]
I used DQN for it. I used the default parameters of dqn_atari.py

**Comment by guptavaibhav18197** [comment]
I used DQN for it, with the default parameters of dqn_atari.py

**Comment by mirraaj** [comment]
Will have a look at it as soon as I get time.

**Comment by guptavaibhav18197** [comment]
@mirraaj I ran the same code on Atlantis-v0. It's not working there too. I saw that someone has reported similar issue for Enduro-v0 #83  too. It would be great if this could be looked into - it would be enormous help of many researchers out there. :)

**Comment by mirraaj** [comment]
@guptavaibhav18197 I guess the problem lies in exploration vs exploitation. When the reward is sparse, exploration becomes a necessity to get the best approximate of Q values. I am looking into the code base to check if there is a problem in the framework. Could you please let me know how long did you run the code and the total number of steps for which the code ran.

**Comment by mirraaj** [comment]
I guess you should start with ```Boxing-ram-v0``` and ```Atlantis-ram-v0``` to begin your experiments. These seem to be an easier task than the one you are trying. I do not see any problem with Boxing results. You can expect such decrease in performance once the DQN network starts overestimation. However, I am not convinced by the Atlantis results. Your graph shows that there is no improvement in the DQN network altogether. I am particularly not sure why this is happening. @matthiasplappert Could you please give your input :)

**Comment by albarji** [comment]
Hi there! I'm also experiencing similar results in `Pong-v0`, using dqn_atari.py . I have also tried modifying the script a little bit to follow more closely the experimental setup of the Nature paper, by using RMSprop(lr=.00025, rho=0.95, epsilon=0.01) instead of Adam. Also added BatchNormalization after each Conv2D+ReLU block, hoping it would help in stabilizing the training process. 

Unfortunately after 5 million steps I can't see any progress.

![imagen](https://user-images.githubusercontent.com/9654655/41761721-a7615044-75f8-11e8-9788-324aad7ee65f.png)

Is this the expected result? The only feature mentioned in the Nature paper which I did not see implemented in dqn_atari.py (besides RMSProp) is some preprocess in which the value of each pixel in a frame is taken as the max between itself and the same pixel in the previous frame. This is reported to be useful to prevent Atari issues with flickering sprites. Could this be the reason? Am I missing some other critical point?

**Comment by wangjksjtu** [comment]
Yeah, I face the same issue. It seems that the network learns nothing. I just use the dqn_atari.py code in examples/ folder.

**Comment by mirraaj** [comment]
@wangjksjtu, Do you face the same issue when dealing with simpler environments like CartPole or MountainCar?

**Comment by mirraaj** [comment]
@wangjksjtu I will be looking into this issue in the coming weekend. Please let me know if you have any intuition what should be wrong. Also, please post the results for atari, if you have it.

**Comment by wangjksjtu** [comment]
@mirraaj, thanks a lot! If the code works for you, could you please share the settings or the code?
I trained the models with 1e7 steps. The results  ([pong.pdf](https://github.com/keras-rl/keras-rl/files/2300219/pong.pdf)) are plotted by visualize_log.py script in examples/ folder. From episode rewards, we could know the model learned nothing.

**Comment by mirraaj** [comment]
@wangjksjtu Could you share your codes? I might want to have a look at it.

**Comment by wangjksjtu** [comment]
@mirraaj I just modify the env to PongNoFrameskip-v4 (dqn_atari.py in examples/ subfolder).

**Comment by wangjksjtu** [comment]
@mirraaj The code is as follows. Thanks a lot!
```
from __future__ import division
import argparse
import os

from PIL import Image
import numpy as np
import gym

import tensorflow as tf
from keras.models import Sequential
from keras.layers import Dense, Activation, Flatten, Convolution2D, Permute
from keras.optimizers import Adam
import keras.backend as K

from rl.agents.dqn import DQNAgent
from rl.policy import LinearAnnealedPolicy, BoltzmannQPolicy, EpsGreedyQPolicy
from rl.memory import SequentialMemory
from rl.core import Processor
from rl.callbacks import FileLogger, ModelIntervalCheckpoint

parser = argparse.ArgumentParser()
parser.add_argument('--log_dir', default='logs',
                    help='Log dir [default: logs]')
parser.add_argument('--mode', choices=['train', 'test'], default='train', 
                    help='Mode (train/test) [default: train]')
parser.add_argument('--env-name', type=str, default='PongNoFrameskip-v4',
                    help='Atari Game [default: PongNoFrameskip-v4]')
parser.add_argument('--pre_trained', type=str, default=None,
                    help='The path of pre-trained weights [default: None]')
FLAGS = parser.parse_args()

MODE = FLAGS.mode
ENV_NAME = FLAGS.env_name
WEIGHTS = FLAGS.pre_trained
LOG_DIR = FLAGS.log_dir

INPUT_SHAPE = (84, 84)
WINDOW_LENGTH = 4

if not os.path.exists(LOG_DIR):
    os.makedirs(LOG_DIR)
os.system('cp dqn_atari.py %s' % (LOG_DIR)) # bkp of train procedure


class AtariProcessor(Processor):
    def process_observation(self, observation):
        assert observation.ndim == 3  # (height, width, channel)
        img = Image.fromarray(observation)
        img = img.resize(INPUT_SHAPE).convert('L')  # resize and convert to grayscale
        processed_observation = np.array(img)
        assert processed_observation.shape == INPUT_SHAPE
        return processed_observation.astype('uint8')  # saves storage in experience memory

    def process_state_batch(self, batch):
        # We could perform this processing step in `process_observation`. In this case, however,
        # we would need to store a `float32` array instead, which is 4x more memory intensive than
        # an `uint8` array. This matters if we store 1M observations.
        processed_batch = batch.astype('float32') / 255.
        return processed_batch

def train():
    # Get the environment and extract the number of actions.
    env = gym.make(ENV_NAME)
    np.random.seed(123)
    env.seed(123)
    nb_actions = env.action_space.n

    config = tf.ConfigProto()
    config.gpu_options.allow_growth = True
    sess = tf.Session(config=config)
    from keras import backend as K
    K.set_session(sess)

    # Next, we build our model. We use the same model that was described by Mnih et al. (2015).
    input_shape = (WINDOW_LENGTH,) + INPUT_SHAPE
    model = Sequential()
    if K.image_dim_ordering() == 'tf':
        # (width, height, channels)
        model.add(Permute((2, 3, 1), input_shape=input_shape))
    elif K.image_dim_ordering() == 'th':
        # (channels, width, height)
        model.add(Permute((1, 2, 3), input_shape=input_shape))
    else:
        raise RuntimeError('Unknown image_dim_ordering.')
    model.add(Convolution2D(32, 8, 8, subsample=(4, 4)))
    model.add(Activation('relu'))
    model.add(Convolution2D(64, 4, 4, subsample=(2, 2)))
    model.add(Activation('relu'))
    model.add(Convolution2D(64, 3, 3, subsample=(1, 1)))
    model.add(Activation('relu'))
    model.add(Flatten())
    model.add(Dense(512))
    model.add(Activation('relu'))
    model.add(Dense(nb_actions))
    model.add(Activation('linear'))
    model.summary()


    memory = SequentialMemory(limit=1000000, window_length=WINDOW_LENGTH)
    policy = LinearAnnealedPolicy(EpsGreedyQPolicy(), attr='eps', value_max=1., value_min=.1, value_test=.05,
                                  nb_steps=1000000)
    # policy = BoltzmannQPolicy(tau=1.)
    
    processor = AtariProcessor()
    dqn = DQNAgent(model=model, nb_actions=nb_actions, policy=policy, memory=memory,
		   processor=processor, nb_steps_warmup=50000, gamma=.99, target_model_update=10000,
		   train_interval=4, delta_clip=1.)
    dqn.compile(Adam(lr=.00025), metrics=['mae'])

    if MODE == 'train':
	weights_filename = os.path.join(LOG_DIR, 'dqn_{}_weights.h5f'.format(ENV_NAME))
	checkpoint_weights_filename = os.path.join(LOG_DIR, 'dqn_normal_' + ENV_NAME + '_weights_{step}.h5f')
	log_filename = os.path.join(LOG_DIR, 'dqn_{}_log.json'.format(ENV_NAME))
	callbacks = [ModelIntervalCheckpoint(checkpoint_weights_filename, interval=250000)]
	callbacks += [FileLogger(log_filename, interval=100)]
	dqn.fit(env, callbacks=callbacks, nb_steps=20000000, log_interval=10000, verbose=2)

	dqn.save_weights(weights_filename, overwrite=True)
	dqn.test(env, nb_episodes=10, visualize=False, verbose=2)

    elif MODE == 'test':
	weights_filename = os.path.join(LOG_DIR, 'dqn_{}_weights.h5f'.format(ENV_NAME))
	if WEIGHTS:
	    weights_filename = WEIGHTS
	    dqn.load_weights(weights_filename)
	    dqn.test(env, nb_episodes=10, visualize=False)


if __name__ == "__main__":
    train()
```

**Comment by wangjksjtu** [comment]
@mirraaj  Hi, do you figure out why it happens? Could you achieve convergence in you exps? Thank you.
